Add wxDPIChangedEvent::Scale() and use it

vadz opened the pull request:
This is just a code simplification which shouldn't result in any real changes.

@MaartenBent could you please have a look at this, notably at wxSlider code? I don't understand why do we use `ceil()` and `floor()` there but nowhere else. Perhaps we should use them in the new `Scale()` methods too?

vadz committed:
Document wxMulDivInt32() and add a test for it

This function exists since always and is probably already used outside
wx, so make it officially public and add at least a trivial unit test
for it.

vadz committed:
Add wxRescaleCoord() helper and use it

No real changes, just make the code a bit shorter and more clear by
using a dedicated function rather than wxMulDivInt32() directly.

vadz committed:
Add wxDPIChangedEvent::Scale() and use it in this event handlers

This is more concise and less error-prone than multiplying/dividing DPI
values manually.

No real changes except, maybe, in wxSlider code where the rounding was
done differently before for some reason.

MaartenBent commented:
> @MaartenBent could you please have a look at this, notably at wxSlider code? I don't understand why do we use `ceil()` and `floor()` there but nowhere else. Perhaps we should use them in the new `Scale()` methods too?

The ceil/floor are used so the thumb gets a better approximation of the expected size.
When starting an application with a slider at 100% the initial thumb is `22px`, when starting at 150% the thumb is `34px`, and when starting at 200% the thumb is `45px`.

When moving the app from 150% to 100% it becomes `34*(96/144)=22,6`. With the old method this results in `floor(22,6)=22px`. with the new method `round(22,6)=23px`.
When moving back to 150% the old method results in `22*(144/96)=floor(33.0)=33px`. While the new method results in `23*(144/96)=round(34,5)=35px`.
A similar thing happens when moving between 200% and 100%. With the old method `45px` becomes `22px` and then `44px`. With the new method it becomes `23px` and then `46px`.

Note that both methods are incorrect. I just used the first one because it always results in `22px` at 100%, which matches the size when starting the app at 100%.
I don't have objections to using the new method.

MaartenBent commented:
Note that we use the same ceil/floor approach in `ScaleCoordIfSet` when changing the size of borders, sizers and windows when the DPI changes.
https://github.com/wxWidgets/wxWidgets/blob/9028ae74fcdd020224ddcb3a3e98519608ca57b7/src/msw/window.cpp#L4901

So maybe it is better to use this in the `wxDPIChangedEvent` scale functions too.

vadz commented:
Thanks for the explanations and for the reminder about `ScaleCoordIfSet()` -- I thought we already had something like this somewhere, but forgot where and didn't find it myself.

I definitely think we should use the same scaling in both cases, and I hoped that using `MulDiv()` would be closer to what Windows itself does, but as, apparently, I was wrong about this, we could also use the alternative approach -- but it should be the same in any case. And using `MulDiv()` still seems simpler, so I guess we should go with it?

There is another bothersome inconsistency: I've used `new, old` order in the new functions while the existing `MSWUpdateOnDPIChange()` uses `old, new` order. I guess `wxRescaleCoord()` should be changed to use the same order too...

Finally, thinking more about this, it still seems too easy to mix up these 2 parameters, so maybe we need something more type-safe here? Or is it not worth it, considering that most of the future uses should be via `wxDPIChangedEvent::Scale()`, where there is no danger of mixing anything up?

PBfordev commented on include/wx/event.h:
```diff
@@ -3140,6 +3140,13 @@ class WXDLLIMPEXP_CORE wxDPIChangedEvent : public wxEvent
     wxSize GetOldDPI() const { return m_oldDPI; }
     wxSize GetNewDPI() const { return m_newDPI; }
 
+    // Scale the value by the radio between new and old DPIs carried by this
```
```suggestion
    // Scale the value by the ratio between new and old DPIs carried by this
```

vadz replied:
Oops, thanks for noticing!

vadz committed:
Make wxRescaleCoord() private functions

They're probably not that useful in application code, which should just
use wxDPIChangedEvent::Scale() instead, so move them to a private
header instead of making them part of the public API.

vadz committed:
Make wxRescaleCoord() safer by requiring explicitly using From/To

This should help with accidentally exchanging the order of parameters
and makes the code using this function more readable.

No real changes.

vadz committed:
Use wxRescaleCoord() in wxMSW DPI update code too

Use the same scaling function as elsewhere instead of using
ceil/floor().

Provide wxRescaleCoordWithFrom specialization for int to allow using it
with just single coordinates too.

vadz commented:
So I've implemented my more type-safe idea, but now I'm not sure if it's worth the extra complexity. What do you think?

In any case, I've decided to keep it private for now.

MaartenBent commented on src/common/wincmn.cpp:
```diff
@@ -2907,10 +2904,7 @@ wxWindowBase::FromDIP(const wxSize& sz, const wxWindowBase* w)
-    // "unspecified" which should be preserved.
-    return wxSize(sz.x == -1 ? -1 : wxMulDivInt32(sz.x, dpi.x, baseline),
-                  sz.y == -1 ? -1 : wxMulDivInt32(sz.y, dpi.y, baseline));
+    return wxRescaleCoord(sz).From(baseline, baseline).To(dpi);
```
You could use `GetStdPPI` instead of `GetStdPPIValue()`.

vadz replied:
Thanks, changed.

MaartenBent commented on src/common/wincmn.cpp:
```diff
@@ -2921,10 +2915,7 @@ wxWindowBase::ToDIP(const wxSize& sz, const wxWindowBase* w)
-    // "unspecified" which should be preserved.
-    return wxSize(sz.x == -1 ? -1 : wxMulDivInt32(sz.x, baseline, dpi.x),
-                  sz.y == -1 ? -1 : wxMulDivInt32(sz.y, baseline, dpi.y));
+    return wxRescaleCoord(sz).From(dpi).To(baseline, baseline);
```
Same as above.

MaartenBent commented on include/wx/private/rescale.h:
```diff
@@ -0,0 +1,145 @@
+    {
+        T value(m_value);
+
+        if ( value.x != -1 )
```
Is `wxDefaultCoord` instead of `-1` useful in this file?

vadz replied:
I guess we should indeed use it. Personally I'm not really sure if `wxDefaultCoord` is more readable than `-1` but we did use it in `ScaleCoordIfSet()`, so there doesn't seem to be any reason not to use it here... I'll update the code, thanks.

MaartenBent reviewed (commented):
Looks good to me.
